.github/workflows/test.yml

```diff
@@ -0,0 +1,63 @@
+name: Main Test Job
+
+on:
+  workflow_call:
+    inputs:
+      mock-maker:
+        required: true
+        type: string
+      member-accessor:
+        required: true
+        type: string
+      java:
+        required: true
+        type: number
+
+jobs:
+  #
+  # Main test job
+  #
+  test:
+    runs-on: ubuntu-latest
+    if: "! contains(toJSON(github.event.commits.*.message), '[skip ci]')"
+
+    # All test steps
+    steps:
+      - name: 1. Checkout out test code
+        uses: actions/checkout@v3
+
+      - name: 2. Download build artifacts with name ${{ inputs.mock-maker }}-${{ inputs.member-accessor }}-build-artifacts
+        uses: actions/download-artifact@v3
+        with:
+          name: ${{ inputs.mock-maker }}-${{ inputs.member-accessor }}-build-artifacts
+
+      - name: 3. Set up Java ${{ inputs.java }}
+        uses: actions/setup-java@v3
+        with:
+          distribution: 'zulu'
+          java-version: ${{ inputs.java }}
+
+      - name: 4. Test on Java ${{ inputs.java }} with ${{ inputs.mock-maker }} and ${{ inputs.member-accessor }}
+        run: |
+          ./gradlew test -x testClasses \
+          -x compileTestJava -x compileTestGroovy -x compileTestKotlin -x compileTestBundleJava \
+          -x processTestResources \
+          -x classes -x compileJava -x compileGroovy -x compileKotlin -x compileOtherBundleJava \
+          -x processResources -x processOtherBundleResources \
+          -x preBuild -x preDebugBuild \
+          -x jar
```

Comment:
This looks risky, if you miss a task (e.g. by project evolution), it might run a mixed build and there might be no visible effect. Also as far as I remember this might fail in Gradle 8.

I think instead of uploading the build directory, it might be a more stable approach to run `publishToAllPublicationsMavenRepository` (see https://github.com/mockito/mockito/blob/main/gradle/java-publication.gradle#L100) and only save that folder with the "production-like" jars. The question is how it would be used, and the answer might be [Gradle dependency substitution](https://docs.gradle.org/current/userguide/resolution_rules.html) done conditionally.

Reply:
agree that it is risky, not sure if my approach was right
any ideas on how to only execute test aka compile once test all? as gradlew test still runs build and compile tasks due to dependent tasks

based on the link, the jar files are published to a local folder, hence they cannot be shared between jobs. also compiled classes and jars are needed together so that compiled classes are used to run the test aka compile once test all. is the idea to upload compiled classes and jars in a remote maven repo to be shared between jobs?

Reply:
Based on my current knowledge there are two ways to go about this:
1. substitute the `main` sourceSet of each project with a pre-built .jar file (based on a `-P` flag), unsure if this is actually feasible. Gradle has a pretty powerful `configuration` and `dependencies` API, so it should be possible.
2. Set up toolchains and create multiple test tasks (see how [Gradle Stutter plugin creates multiple test tasks for each Java * Gradle version](https://github.com/gradle-nexus/publish-plugin/actions/runs/5181270374/jobs/9336551191#step:4:175)).

The second one would create multiple builds (because of the maintainer's requirements to keep the matrix), but the .jar files would be exactly the same. As far as I read it, in the referenced blog post this is not a problem because there's no matrix and the toolchain approach is used.

The first approach would optimize the build pipeline for matrix, but I've never seen an example like this. At my previous company with did a similar `-x` hack for a while, but it fell apart during Gradle/Android Gradle Plugin upgrades. Dependency substitution is a less-hacky way of achieving what `-x` does.

Reply:
the second approach creates different gradle task on the same runner/gh job. From the response in the original issue, it seems different gh job is needed to test for each java version

Reply:
Yeah, that works pretty well with matrix: see [how to dynamically use a task](https://github.com/gradle-nexus/publish-plugin/blob/master/.github/workflows/java-versions.yml#L40).

Reply:
Sorry, I am still trying to understand this. Starting with build artifacts, instead of `upload-artifacts`, use `publishToAllPublicationsMavenRepository`.

1) this uploads artifacts into a local maven repo. How are these artifacts shared between github jobs? as local repo is local to each runner
2) jar + compiled classes is needed for testing. So is it advisable to upload the entire build folder?